### ISSUE TITLE
Optimization request: initialization of CardView lacks feedback

### ISSUE THREAD

**Issue by emyoulation** [cdhorn/CardView#192]
The 1st time a CardView mode is activated after loading a large tree, there is significant delay before the Gramps app begins a redraw. 

This gives the impression that the mode change has been ignored. Impatient users may select another mode instead of allowing the view to initialize.

There should be some sort of indication that "stuff" is being populated. Maybe the Statusbar could have a initializing message or a progress-bar to avoid the "not responding" automatically appearing in the Titlebar after 5 seconds. (Is that related to `gsettings set org.gnome.mutter check-alive-timeout` ?)

Ideally, maybe threaded CardView housekeeping task could be assigned to an idle core of a multi-core processor when a new tree is loaded? The high overhead could be pre-processed on a low priority and that initial delay would disappear unless a CardView was the active view mode during a tree load.

Note:
Dunno if this [stackoverflow thread](https://stackoverflow.com/questions/4886634/python-gtk-window-not-responding) answer has any bearing:
You can use [gobject.idle_add](http://www.pygtk.org/pygtk2reference/gobject-functions.html#function-gobject--idle-add)  to have a function be executed and still keep your windows responsive

**Comment by cdhorn** [comment]
Without LastChanged Gramplet? That one even with yeild and large trees is problematic.

The Gramplet rendering look and I think status bar updates leverage idle_add as does my deferred config update logic.

Really need https://github.com/gramps-project/gramps/pull/1119 first to think about how might use threads and if makes sense as Gramps was never designed with that in mind.

**Comment by fxtmtrsine** [comment]
> The 1st time a CardView mode is activated after loading a large tree, there is significant delay before the Gramps app begins a redraw.

Even with the fix from #184 (thank you) I ended up removing the Gramplet permanently (was waiting 17 seconds between view changes!), my suggestion is don't install it by default to give a better experience overall for people with larger trees?

Is it possible to cache the results between Gramps sessions ( exiting totally and opening/closing different family trees ) as the Gramplet seems to rescans each time?

**Comment by cdhorn** [comment]
> Not only is Gramps not designed with threading in mind, neither was Python.

The GIL limits things, but there are lots of scenarios where it is more than fine. It all depends on what you need to do.

Lots of big projects reach certain points where the weight of the technical debt needs to be considered and how best to deal with it. Do you keep trudging along?  Do you try to refactor enough of it to keep things moving forward without breaking compatibility? Or is it time to set the codebase aside, although maybe cherry picking it, to write something new from scratch taking all the lessons learned with you to start with a more solid foundation?

If the Gramps team was to do it all over again, with all the existing experience behind them, what would everyone do differently?

> my suggestion is don't install it by default

Yes I agree I should probably remove it as a default and will soon.  It can be useful, but with large trees it can be pretty painful on startup or if it needs to rescan.

It would be possible to stash the results and pull them in again, but then one has to worry about stuff getting out of sync with the db and not being sure although I do force rescan on delete if I recall. Maybe with last db change time can work around some of that uncertainty and if I kept a deeper history I'd not worry about rescan on delete.  If time permits I might look at it.

**Comment by emyoulation** [comment]
The dashboard locked up Gramps for 76 seconds while gathering statistics when switching trees from the Example.gramps tree to my 45,000 person tree. During the time it was gathering data, the stats from the previous tree remained visible.

The GUI was completely unavailable. I could not interrupt to switch to another view or even turn off the sidebar/Bottombar.

It needs so interrupts so it plays nice. Or at least a 'busy'.

Maybe it could zero the stats before starting to gather? 

Another progressbar-like option might be to make the tile section headers all have red backgrounds before their stats are gathered. And, as each section is gathered, don't populate the values, merely change the background to normal.

**Comment by emyoulation** [comment]
This session:
render_page: dashboard 54.49921631813049
![image](https://user-images.githubusercontent.com/69127217/174342159-15a4d407-6df3-4a3d-85ef-7635a568d81b.png)

**Comment by cdhorn** [comment]
Yeah that's the #199 issue. I need to spin up the Windows VM to look at that at some point.  It isn't happening for me on Fedora.

**Comment by PQYPLZXHGF** [comment]
> Yes I am painfully aware it is an issue, no different than the last changed Gramplet.
> 
> I am spending a lot of time thinking about how to best handle this.

I know it is not a Gramplet but maybe consider doing something similar to how [minimizing a Gramplet](https://www.gramps-project.org/wiki/index.php/Gramplets#Advanced_Settings) will prevent it from [updating](https://github.com/gramps-project/gramps/blob/152d24fad1de13965bc81038fff549cb86865808/gramps/plugins/gramplet/sessionloggramplet.py#L49).  (**force_update**: set **True** to have the gramplet update, even when minimized)  and then by default have almost all of the statistics sections closed and users could show all or hide all like others mention and have card view remember between sessions what the user wants? So opening the section would start the information gathering of that type and you can have a timer icon or word message showing its in progress?

**Comment by emyoulation** [comment]
I think that the big question is: even if the Gramplet takes a bit longer on smaller trees, does the main window gain in responsiveness? The latency due to overhead of the Gramplets tends to make the main view unusable if it is processor intensive too.

**Comment by cdhorn** [comment]
This is the view, not the Gramplet, but in both cases it is the same issue.

Threading is a natural way to deal with the UI responsiveness. A thread is dispatched to do the work and the main UI thread can continue to run it's event loop without blocking. 
 
In all of Gramps there is only a single peice of code that runs in a separate thread in gramps.gui.utils:
```
class AvailableUpdates(threading.Thread):
    def __init__(self, uistate):
        threading.Thread.__init__(self)
        self.uistate = uistate
        self.addon_update_list = []

    def emit_update_available(self):
        self.uistate.emit('update-available', (self.addon_update_list, ))

    def run(self):
        self.addon_update_list = available_updates()
        if len(self.addon_update_list) > 0:
            GLib.idle_add(self.emit_update_available)
```
I guess this is probably because of all the technical debt around the Berkley database.

Note in my large database that takes 12 seconds on average to render I ran some experiments.  I found it took 5-5.5 seconds to read the data in from the database and instantiate the objects without doing any actual work. It took 1.2 seconds to read the raw objects in, but that includes unpickling the serialized data. When I examined that I found it took .8 seconds to either pickle or unpickle the whole data set, so it actually only took .4 seconds to read through the entire database.

So almost half the processing time is spent doing nothing meaningful with the actual data.

**Comment by cdhorn** [comment]
I guess this subprocess/multiprocessing approach is not going to work with the Windows AIO package. It seems to use a launcher of some sort and doesn't actually have a python executible I can call to run something externally.

**Comment by emyoulation** [comment]
Well, it was a slim hope and thanks for exploring it.

Hopefully it works for your Linux installation though?

**Comment by cdhorn** [comment]
Collection is done in a separate thread now and cards updated when statistics are ready.  Quick test on Windows VM it seems to work okay.

There is much left to be done with this. ie refreshing stats, handling close db properly when thread collecting data, update last changed gramplet, and stuff like that.
